### PR TITLE
[Datasets] Remove "ML Preprocessing" user guide

### LINKED ISSUE #30401: Couldn't see the checkpoint files from Ray rllib 2.1.0
### What happened + What you expected to happen

From Rllib 2.1.0, I am using Marwil model on Custom Environment and using python Trainer API , I couldn't see Checkpoint files anymore, I am using agent.save(), and when trying to do the restore checkpoints, Its has some other files like meta files for .is_checkpoint, .tune_metadata, algorithm_state.pkl, rllib_checkpoint and policies Directory, I tried load pkl file. but its not working !!!

But when I am using Rllib 2.0.0, I can see the Checkpoint files !!!

[checkpoints](https://user-images.githubusercontent.com/45755237/202391512-ebd4ed16-06aa-4b00-b86d-1691443419a8.png)

### Versions / Dependencies

python 3.8.8 
ray rllib 2.1.0 

### Reproduction script

train_start = time.process_time()
MAX_EPISODES = 601
print("Training started !!!")
for i in range(1,MAX_EPISODES):
    start = time.process_time()
    print(f"Episode no: {i}")
    print()
    results = trainer.train()
    print(f"Time taken for Episode-{i} : ",time.process_time() - start)
    if i % 100 == 0:
        check_points = trainer.save('checkpoint/marwil_new')
        print("Checkpoint saved!!!")

### Issue Severity

_No response_

### LINKED ISSUE #30410: [Datasets] Revise "ML Preprocessing" page
### Description

There're a few issues:
* No mention of AIR preprocessors (this is the big one)
* Overlap with "Transforming datasets"
* Long code snipppets

### Link

https://docs.ray.io/en/master/data/dataset-ml-preprocessing.html

### PR DESCRIPTION
**Pull request by bveeramani**
## Why are these changes needed?

The "ML Preprocessing" guide contains outdated or redundant information. With the introduction of the "Using Preprocessors" and "Configuring Training Datasets" guides, the page is largely obsolete. This PR removes "ML Preprocessing" and moves still-relevant information to other guides.

## Related issue number

Closes https://github.com/ray-project/ray/issues/30410

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/data/dataset.rst [hunk 1]
```diff
@@ -36,8 +36,22 @@ Data Loading and Preprocessing for ML Training
 Ray Datasets are designed to load and preprocess data for distributed :ref:`ML training pipelines <train-docs>`.
 Compared to other loading solutions, Datasets are more flexible (e.g., can express higher-quality `per-epoch global shuffles <examples/big_data_ingestion.html>`__) and provides `higher overall performance <https://www.anyscale.com/blog/why-third-generation-ml-platforms-are-more-performant>`__.
 
-Ray Datasets are not intended as a replacement for more general data processing systems.
-:ref:`Learn more about how Ray Datasets work with other ETL systems <datasets-ml-preprocessing>`.
+Datasets is not intended as a replacement for more general data processing systems.
+Its utility is as the last-mile bridge from ETL pipeline outputs to distributed applications and libraries in Ray:
```
Comment (angelinalg) [inline doc/source/data/dataset.rst:40]:
```suggestion

```

Comment (angelinalg) [reply doc/source/data/dataset.rst:40]: If you're comfortable being more opinionated about this guidance. Think of how you would tell a friend who is using Datasets.

Also, do we need to explain what we mean by `last-mile`? I thought that might be a term that we coined internally. If not, no need to explain.

File: doc/source/data/dataset.rst [hunk 2]
```diff
@@ -36,8 +36,22 @@ Data Loading and Preprocessing for ML Training
...
+
+.. image:: images/dataset-loading-1.png
+   :width: 650px
+   :align: center
+
+..
+  https://docs.google.com/presentation/d/1l03C1-4jsujvEFZUM4JVNy8Ju8jnY5Lc_3q7MBWi2PQ/edit
+
+Ray-integrated DataFrame libraries can also be seamlessly used with Datasets, to enable running a full data to
```
Comment (angelinalg) [inline doc/source/data/dataset.rst:49]:
```suggestion
You can also use Datasets seamlessly with Ray-integrated DataFrame libraries. You can run a data to
```

File: doc/source/data/dataset.rst [hunk 3]
```diff
@@ -36,8 +36,22 @@ Data Loading and Preprocessing for ML Training
...
+ML pipeline completely within Ray without requiring data to be materialized to external storage:
```
Comment (angelinalg) [inline doc/source/data/dataset.rst:50]:
```suggestion
ML pipeline completely within Ray without materializing data to external storage:
```

Comment (angelinalg) [reply doc/source/data/dataset.rst:50]: Not sure if I mangled the intended meaning, but I'm trying to avoid some passive voice.

File: doc/source/data/transforming-datasets.rst [hunk 4]
```diff
@@ -113,8 +113,8 @@ There are three types of UDFs that you can use with Ray Data: Function UDFs, Cal
   When using the actor compute strategy, per-row and per-batch UDFs can also be
```
Comment (angelinalg) [inline doc/source/data/transforming-datasets.rst:113]:
```suggestion
  With the actor compute strategy, you can use per-row and per-batch UDFs
```

File: doc/source/data/transforming-datasets.rst [hunk 5]
```diff
@@ -113,8 +113,8 @@ There are three types of UDFs that you can use with Ray Data: Function UDFs, Cal
...
   *callable classes*, i.e. classes that implement the ``__call__`` magic method. The
```
Comment (angelinalg) [inline doc/source/data/transforming-datasets.rst:114]:
```suggestion
  *callable classes*, i.e., classes that implement the ``__call__`` magic method. The
```

File: doc/source/data/transforming-datasets.rst [hunk 6]
```diff
@@ -113,8 +113,8 @@ There are three types of UDFs that you can use with Ray Data: Function UDFs, Cal
...
   constructor of the class can be used for stateful setup, and will be only invoked once
```
Comment (angelinalg) [inline doc/source/data/transforming-datasets.rst:115]:
```suggestion
  can use the constructor of the class for stateful setup, and it is only invoked once
```

File: doc/source/data/transforming-datasets.rst [hunk 7]
```diff
@@ -113,8 +113,8 @@ There are three types of UDFs that you can use with Ray Data: Function UDFs, Cal
...
-  per worker actor. 
-  
+  per worker actor.
+
   Callable classes are useful if there is expensive state (such as a neural network) that need to be loaded for the UDF. By using an actor class, the state only needs to be loaded once in the beginning, rather than for each batch.
```
Comment (angelinalg) [inline doc/source/data/transforming-datasets.rst:118]:
```suggestion
  Callable classes are useful if you need to load an expensive state (such as a neural network) for the UDF. By using an actor class, the you only need to load the state once in the beginning, rather than for each batch.
```

File: doc/source/data/dataset.rst [hunk 8]
```diff
@@ -36,8 +36,22 @@ Data Loading and Preprocessing for ML Training
 Ray Datasets are designed to load and preprocess data for distributed :ref:`ML training pipelines <train-docs>`.
```
Comment (angelinalg) [inline doc/source/data/dataset.rst:36]:
```suggestion
Use Ray Datasets to load and preprocess data for distributed :ref:`ML training pipelines <train-docs>`.
```

File: doc/source/data/dataset.rst [hunk 9]
```diff
@@ -36,8 +36,22 @@ Data Loading and Preprocessing for ML Training
...
 Compared to other loading solutions, Datasets are more flexible (e.g., can express higher-quality `per-epoch global shuffles <examples/big_data_ingestion.html>`__) and provides `higher overall performance <https://www.anyscale.com/blog/why-third-generation-ml-platforms-are-more-performant>`__.
 
-Ray Datasets are not intended as a replacement for more general data processing systems.
-:ref:`Learn more about how Ray Datasets work with other ETL systems <datasets-ml-preprocessing>`.
+Datasets is not intended as a replacement for more general data processing systems.
```
Comment (angelinalg) [inline doc/source/data/dataset.rst:39]:
```suggestion
Use Datasets for the last-mile bridge from ETL pipeline outputs to distributed applications and libraries in Ray. Do not use them as a replacement for more general data processing systems.
```

**Comment by angelinalg** [comment]
Very excited about the maintenance work you're doing on the docs!

File: doc/source/data/dataset.rst (hunk 3 above)
Comment (angelinalg) [inline doc/source/data/dataset.rst:50]:
```suggestion
ML pipeline completely within Ray without requiring data to be materialized to external storage.
```

File: doc/source/data/dataset.rst (hunk 1 above)
Comment (bveeramani) [reply doc/source/data/dataset.rst:40]: Hmm...really not sure how I'd describe Datasets' utility in a simple but precise way.

Maybe something like "Datasets moves data from storage to distributed Ray applications. It also lets you transform data into representations that are more suitable for models."?

**Comment by amogkam** [comment]
Thanks @bveeramani. 

1. the random shuffling/randomize block order section is being dropped right? Where should I be looking at for this information?
2. Can we create more synergy with the AIR docs? can we link to the AIR preprocessing pages and Check ingest pages as appropriate? For example, refer to the AIR preprocessing doc in the "Transforming Datasets" section

**Comment by bveeramani** [comment]
> 1. the random shuffling/randomize block order section is being dropped  in this PR? Where should I be looking at for this information?

[Shuffling Data](https://docs.ray.io/en/latest/ray-air/check-ingest.html#shuffling-data).

> 2. Can we create more synergy with the AIR docs? can we link to the AIR preprocessing pages and Check ingest pages as appropriate? For example, refer to the AIR preprocessing doc in the "Transforming Datasets" section

Yeah, will add.

**Comment by amogkam** [comment]
It's not clear I need to go to AIR docs to see how to do shuffling

File: doc/source/ray-air/check-ingest.rst [hunk 10]
```diff
@@ -131,7 +131,13 @@ Enabling Streaming Ingest
 Shuffling Data
 ~~~~~~~~~~~~~~
 
-Shuffling or data randomization is important for training high-quality models. By default, AIR will randomize the order the data files (blocks) are read from. AIR also offers options for further randomizing data records within each file:
+Shuffling or data randomization is important for training high-quality models.
+
+By default, AIR reads data files (blocks) in a random order and shuffles the blocks
+assigned to each dataset shard. You can disable this behaviour by setting
```
Comment (clarkzinzow) [inline doc/source/ray-air/check-ingest.rst:137]:
```suggestion
By default, AIR shuffles the assignment of data blocks (files) to dataset shards between epochs. You can disable this behavior by setting
```

Comment (clarkzinzow) [reply doc/source/ray-air/check-ingest.rst:137]: Also not sure if we should call out how to disable this, since this is a very cheap operation that is an easy win for improving training data shuffling. Is there a use case in which you'd want to disable this? If not, I don't think we need a disabling call out.

File: doc/source/data/dataset.rst [hunk 11]
```diff
@@ -33,11 +33,24 @@ is already supported.
 Data Loading and Preprocessing for ML Training
 ----------------------------------------------
 
-Ray Datasets are designed to load and preprocess data for distributed :ref:`ML training pipelines <train-docs>`.
+Use Ray Datasets to load and preprocess data for distributed :ref:`ML training pipelines <train-docs>`.
 Compared to other loading solutions, Datasets are more flexible (e.g., can express higher-quality `per-epoch global shuffles <examples/big_data_ingestion.html>`__) and provides `higher overall performance <https://www.anyscale.com/blog/why-third-generation-ml-platforms-are-more-performant>`__.
 
-Ray Datasets are not intended as a replacement for more general data processing systems.
-:ref:`Learn more about how Ray Datasets work with other ETL systems <datasets-ml-preprocessing>`.
+Use Datasets for the last-mile bridge from ETL pipeline outputs to distributed applications and libraries in Ray. Do not use them as a replacement for more general data processing systems.
```
Comment (clarkzinzow) [inline doc/source/data/dataset.rst:39]:
```suggestion
Use Datasets as a last-mile bridge from storage or ETL pipeline outputs to distributed applications and libraries in Ray. Do not use it as a replacement for more general data processing systems.
```

File: doc/source/data/dataset.rst [hunk 12]
```diff
@@ -33,11 +33,24 @@ is already supported.
...
+   :align: center
+
+..
+  https://docs.google.com/presentation/d/1l03C1-4jsujvEFZUM4JVNy8Ju8jnY5Lc_3q7MBWi2PQ/edit
+
+You can also use Datasets seamlessly with Ray-integrated DataFrame libraries.
+
+.. image:: images/dataset-loading-2.png
+   :width: 650px
+   :align: center
```
Comment (c21) [inline doc/source/data/dataset.rst:53]: I am thinking of to drop this part. If we look at the images, they are mostly identical, except the scope of "Ray cluster". And to run Spark/Mars/etc on Ray is community-maintained, and not battle tested. So it should be safe to drop here.

File: doc/source/data/dataset.rst (hunk 1 above)
Comment (c21) [reply doc/source/data/dataset.rst:40]: Can we remove "Its utility is as the last-mile bridge from ETL pipeline outputs to distributed applications and libraries in Ray:".

This is pretty much repeated from above sentence: "Use Datasets for the last-mile bridge from ETL pipeline outputs to distributed applications and libraries in Ray."

We can add a link to "User Guides" here, e.g. "Check out Datasets User Duides for the features it supports".

**Comment by clarkzinzow** [comment]
+1 to more cross links from Datasets docs to AIR docs.

As for the shuffling docs, why don't we add a shuffle section to the "Transforming Datasets" feature guide that only details the `ds.random_shuffle()` API (basically the section that we have in the current "ML Preprocessing" guide), and have it cross-link to the AIR docs on shuffling + call out that local shuffling is a good alternative, but delegate details to the AIR level docs?

File: doc/source/data/transforming-datasets.rst [hunk 13]
```diff
@@ -373,6 +373,97 @@ The following output types are allowed for per-row UDFs (e.g.,
     :start-after: __writing_simple_out_row_udfs_begin__
     :end-before: __writing_simple_out_row_udfs_end__
 
+.. _datasets-groupbys:
```
Comment (c21) [inline doc/source/data/transforming-datasets.rst:376]: Shall we put this section to the end of this file?

`Writing User-defined Functions (UDFs)` -> `Configuring Batch Size` -> `Compute Strategy` is a coherent story for writing `map_batches`.

File: doc/source/data/transforming-datasets.rst [hunk 14]
```diff
@@ -373,6 +373,97 @@ The following output types are allowed for per-row UDFs (e.g.,
...
+
```
Comment (c21) [inline doc/source/data/transforming-datasets.rst:377]: what about https://docs.ray.io/en/master/data/dataset-ml-preprocessing.html#random-shuffle and https://docs.ray.io/en/master/data/dataset-ml-preprocessing.html#random-block-order? Should we add here as well?

File: doc/source/ray-air/check-ingest.rst (hunk 10 above)
Comment (bveeramani) [reply doc/source/ray-air/check-ingest.rst:137]: If you're evaluating a model on a validation set, you wouldn't want to shuffle data blocks.

File: doc/source/data/dataset.rst (hunk 12 above)
Comment (bveeramani) [reply doc/source/data/dataset.rst:53]: Removed

File: doc/source/data/transforming-datasets.rst (hunk 14 above)
Comment (bveeramani) [reply doc/source/data/transforming-datasets.rst:377]: For `randomize_block_order`, I've added a callout in "Shuffling Data" section of the AIR user guide.

For `random_shuffle`, I'll follow Clark's suggestion:
> As for the shuffling docs, why don't we add a shuffle section to the "Transforming Datasets" feature guide that only details the ds.random_shuffle() API (basically the section that we have in the current "ML Preprocessing" guide), and have it cross-link to the AIR docs on shuffling + call out that local shuffling is a good alternative, but delegate details to the AIR level docs?

**Comment by bveeramani** [comment]
> 1. the random shuffling/randomize block order section is being dropped  in this PR? Where should I be looking at for this information?

Added a "Shuffling data" section to "Transforming data" and linked to the "Shuffling data" section of the AIR user guide.

> 2. Can we create more synergy with the AIR docs? can we link to the AIR preprocessing pages and Check ingest pages as appropriate? For example, refer to the AIR preprocessing doc in the "Transforming Datasets" section

In addition to linking to the AIR docs in "Shuffling data", I also added a reference to the top of "Transforming Datasets" that points to "Using Preprocessors".

**Review by angelinalg — APPROVED** [review]
LGTM

**Review by c21 — APPROVED** [review]
LG
